xformers/triton/ragged_inference/triton_v2_qk_dotprod.py

```diff
@@ -0,0 +1,222 @@
+        triton.Config(
+            {"BLOCK_M": 64, "BLOCK_K": 256, "BLOCK_D": 32},
+            num_stages=4,
+            num_warps=4,
+        ),
+        triton.Config(
+            {"BLOCK_M": 128, "BLOCK_K": 128, "BLOCK_D": 32},
+            num_stages=4,
+            num_warps=4,
+        ),
+        triton.Config(
+            {"BLOCK_M": 128, "BLOCK_K": 64, "BLOCK_D": 32},
+            num_stages=4,
+            num_warps=4,
+        ),
+        triton.Config(
+            {"BLOCK_M": 64, "BLOCK_K": 128, "BLOCK_D": 32},
+            num_stages=4,
+            num_warps=4,
+        ),
+        triton.Config(
+            {"BLOCK_M": 128, "BLOCK_K": 32, "BLOCK_D": 32},
+            num_stages=4,
+            num_warps=4,
+        ),
+        triton.Config(
+            {"BLOCK_M": 64, "BLOCK_K": 32, "BLOCK_D": 32},
+            num_stages=5,
+            num_warps=2,
+        ),
+    ] + get_configs_io_bound()
+
+
+# def get_fast_dev_configs():
+#     return [
+#         triton.Config(
+#             {"BLOCK_Q": 64, "BLOCK_K": 32, "BLOCK_D": 32},
+#             num_stages=5,
+#             num_warps=2,
+#         )
+#     ]
+def get_fast_dev_configs():
+    return [
+        triton.Config(
+            {"BLOCK_Q": 8, "BLOCK_K": 8, "BLOCK_D": 8},
+            num_stages=5,
+            num_warps=2,
+        )
+    ]
+
+
+@triton.autotune(
+    # configs=get_all_configs(),
+    configs=get_fast_dev_configs(),
+    key=["n_ctx_q", "n_ctx_k", "d_model"],
+    prune_configs_by={
+        "prune_num_stages_by": prune_num_stages,
+        "perf_model": estimate_matmul_time,
+        "top_k": 10,
+    },
+)
+@triton.jit
+def _kernel(
+    q_ptr,
+    k_ptr,
+    scores_ptr,
+    n_ctx_q,
+    n_ctx_k,  # N
+    d_model,
+    stride_ctx_q,
+    stride_ctx_k,
+    stride_d,  # Stride along the d_model_per_head dim
+    stride_out_q,
+    stride_out_k,
+    BLOCK_Q: tl.constexpr,
+    BLOCK_K: tl.constexpr,
+    BLOCK_D: tl.constexpr,
+):
+
+    # matrix multiplication
+    pid = tl.program_id(0)
+
+    # Determine the number of blocks in the grid
+    grid_k = (n_ctx_k + BLOCK_K - 1) // BLOCK_K
+
+    pid_q = pid // grid_k
+    pid_k = pid % grid_k
+
+    # do matrix multiplication
+    rq = pid_q * BLOCK_Q + tl.arange(0, BLOCK_Q)
+    rq = tl.max_contiguous(tl.multiple_of(rq % n_ctx_q, BLOCK_Q), BLOCK_Q)
+
+    rk = pid_k * BLOCK_K + tl.arange(0, BLOCK_K)
+    rk = tl.max_contiguous(tl.multiple_of(rk % n_ctx_k, BLOCK_K), BLOCK_K)
+
+    # Iterate through blocks of the d_model dimension and accumulate values into acc
+    acc_tile = tl.zeros((BLOCK_Q, BLOCK_K), dtype=tl.float32)
+    rd = tl.arange(0, BLOCK_D)
+
+    q_ptr_tile = q_ptr + (rq[:, None] * stride_ctx_q + rd[None, :] * stride_d)
+    k_ptr_tile = k_ptr + (rd[:, None] * stride_d + rk[None, :] * stride_ctx_k)
+
+    # acc_tile = tl.load(k_ptr_tile)
+
+    for d_max_offset in range(d_model, 0, -BLOCK_D):
+        q_tile = tl.load(q_ptr_tile, mask=rd[None, :] < d_max_offset, other=0.0)
```

Comment:
(masking) guess here is that BLOCK_Q, BLOCK_K will not tile beyond their respective dimensions (Q and K), which is completely possible, but FYI if you get a mem error on weird sizes that could be why, since this just guards on D. Just mentioning if this helps, not a worry